pumpkin-inventory/src/lib.rs

```diff
@@ -66,29 +67,36 @@ pub trait Container: Sync + Send {
         slot: usize,
         mouse_click: MouseClick,
     ) -> Result<(), InventoryError> {
-        let mut all_slots = self.all_slots();
-        if slot > all_slots.len() {
-            Err(InventoryError::InvalidSlot)?
-        }
-        handle_item_change(carried_item, all_slots[slot], mouse_click);
+        handle_item_change(
+            carried_item,
+            self.get_slot_mut(slot).ok_or(InventoryError::InvalidSlot)?,
+            mouse_click,
+        );
         Ok(())
     }
 
-    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>>;
+    fn iter_slots_mut<'s>(
+        &'s mut self,
+    ) -> Box<dyn ExactSizeIterator<Item = &mut Option<ItemStack>> + 's>;
 
-    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>>;
+    fn iter_slots<'s>(&'s self) -> Box<dyn ExactSizeIterator<Item = &Option<ItemStack>> + 's>;
 
-    fn all_combinable_slots(&self) -> Vec<Option<&ItemStack>> {
-        self.all_slots_ref()
+    fn all_combinable_slots(&self) -> Vec<&Option<ItemStack>> {
+        self.iter_slots().collect_vec()
     }
 
     fn all_combinable_slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
-        self.all_slots()
+        self.iter_slots_mut().collect_vec()
     }
 
     fn internal_pumpkin_id(&self) -> u64 {
         0
     }
+
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>>;
```

Comment:
Maybe put a doc comment here, explaining the difference between the two `Option`